### PR TITLE
EVM transaction replacement

### PR DESCRIPTION
**Pull request by codchen**
## Describe your changes and provide context
Added replacement logic for EVM transactions that works as follows:
- if a tx with the same nonce exists in the priority index that has a lower priority, the existing tx is replaced by the incoming tx
- if a tx with the same nonce exists in the priority index that has an equal or higher priority, the incoming tx is dropped
After this change, there will not be any tx with duplicate nonce in the same queue inside the priority index, so the timestamp check in `IsBefore` is also removed.

As a requirement for keeping the invariant above, for any committed transaction, we will remove any tx in the mempool that has the same nonce regardless of whether it's the same tx or not. (For example, a node may have a tx of nonce-10 with content ABC in its local mempool, but received and executed a block proposal that contains a tx of nonce-10 with content DEF. In this case, we want to remove ABC from the local mempool after DEF's commit)

## Testing performed to validate your change
unit test & local sei test

local sei test steps:
```
Block time -> 10s

block 1
seid tx evm send 0x9f76969569764cFc33D597D0eb00242cDB11EAE2 1000 --from=admin --nonce=0 --gas-fee-cap=1000000000000
seid tx evm send 0x9f76969569764cFc33D597D0eb00242cDB11EAE2 2000 --from=admin --nonce=0 --gas-fee-cap=1100000000000

block 1
seid tx evm send 0x9f76969569764cFc33D597D0eb00242cDB11EAE2 1000 --from=admin --nonce=0 --gas-fee-cap=1000000000000
seid tx evm send 0x9f76969569764cFc33D597D0eb00242cDB11EAE2 2000 --from=admin --nonce=0 --gas-fee-cap=900000000000

block 1
seid tx evm send 0x9f76969569764cFc33D597D0eb00242cDB11EAE2 1000 --from=admin --nonce=0 --gas-fee-cap=1000000000000
seid tx evm send 0x9f76969569764cFc33D597D0eb00242cDB11EAE2 2000 --from=admin --nonce=2 --gas-fee-cap=1000000000000
seid tx evm send 0x9f76969569764cFc33D597D0eb00242cDB11EAE2 3000 --from=admin --nonce=2 --gas-fee-cap=1100000000000
block 2
seid tx evm send 0x9f76969569764cFc33D597D0eb00242cDB11EAE2 10000 --from=admin --nonce=1 --gas-fee-cap=1000000000000
seid tx evm send 0x9f76969569764cFc33D597D0eb00242cDB11EAE2 100000 --from=admin --nonce=3 --gas-fee-cap=1000000000000

block 1
seid tx evm send 0x9f76969569764cFc33D597D0eb00242cDB11EAE2 1000 --from=admin --nonce=0 --gas-fee-cap=1000000000000
seid tx evm send 0x9f76969569764cFc33D597D0eb00242cDB11EAE2 2000 --from=admin --nonce=2 --gas-fee-cap=1000000000000
seid tx evm send 0x9f76969569764cFc33D597D0eb00242cDB11EAE2 3000 --from=admin --nonce=2 --gas-fee-cap=900000000000
block 2
seid tx evm send 0x9f76969569764cFc33D597D0eb00242cDB11EAE2 10000 --from=admin --nonce=1 --gas-fee-cap=1000000000000
seid tx evm send 0x9f76969569764cFc33D597D0eb00242cDB11EAE2 100000 --from=admin --nonce=3 --gas-fee-cap=1000000000000
```

### CODE REVIEW

File: internal/mempool/priority_queue.go [hunk 1]
```diff
@@ -50,6 +53,51 @@ func NewTxPriorityQueue() *TxPriorityQueue {
...
+	}
+	pq.mtx.Lock()
+	defer pq.mtx.Unlock()
+	queue, ok := pq.evmQueue[tx.evmAddress]
+	if ok && len(queue) > 0 {
+		existing, idx := pq.getTxWithSameNonceUnsafe(tx)
+		if existing != nil {
+			if tx.priority > existing.priority {
+				// should replace
+				// replace heap if applicable
```
Comment (philipsu522) [inline internal/mempool/priority_queue.go:86]: nit comment fix

**Review by philipsu522 — APPROVED** [review]
lgtm but let's wait for @stevenlanders to review

File: internal/mempool/mempool.go [hunk 2]
```diff
@@ -364,7 +364,7 @@ func (txmp *TxMempool) RemoveTxByKey(txKey types.TxKey) error {
 
 	// remove the committed transaction from the transaction store and indexes
 	if wtx := txmp.txStore.GetTxByHash(txKey); wtx != nil {
-		txmp.removeTx(wtx, false, true)
+		txmp.removeTx(wtx, false, true, true)
```
Comment (stevenlanders) [inline internal/mempool/mempool.go:367]: (once we have so many flags, it might make sense to split these out as named methods)

Comment (codchen) [reply internal/mempool/mempool.go:367]: yeah makes sense. Or group them into a `RemoveTxOptions` struct that has named fields?

**Pull request merged**